### PR TITLE
Cypress test iterator

### PR DESCRIPTION
**Pull request by dr3**
**Overall change:** Cypress test iterator

**Code changes:**

- _A bullet point list of key code changes that have been made._
- _When describing code changes, try to communicate **how** and **why** you implemented something a specific way, not just **what** has changed._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval

### CODE REVIEW

File: cypress/support/alwaysTests.js
Comment (jamesbhobbs) [inline cypress/support/alwaysTests.js]: I don't think this should be in support. This is effectively a test file. I'd move to /cypress/integration/pages/index.js which would seem logical.

File: cypress/support/alwaysTests.js
Comment (jamesbhobbs) [inline cypress/support/alwaysTests.js]: Ok now I see why testing is taking so long

File: cypress/support/bodyTestHelper.js
Comment (jamesbhobbs) [inline cypress/support/bodyTestHelper.js:12]: What happens if the data requested changes mid test? by testing the JSON sent with a page we avoid this case.

File: cypress/support/commands/pageTestFragment/metaData.js
Comment (jamesbhobbs) [inline cypress/support/commands/pageTestFragment/metaData.js]: This is too simple to make a command, it takes more time to check it under the hood than to read the line. I'd just use cy.title().should('eq', title); everywhere this is used.

File: cypress/support/commands/pageTestFragment/cookie.js
Comment (jamesbhobbs) [inline cypress/support/commands/pageTestFragment/cookie.js]: Again seems too simple just repeat the line.

File: cypress/integration/pages/frontPage.js
Comment (jamesbhobbs) [inline cypress/integration/pages/frontPage.js]: why are we not checking canonical on every page rather than here?

File: cypress/integration/pages/frontPage.js
Comment (jamesbhobbs) [inline cypress/integration/pages/frontPage.js:10]: Why are we testing WS on desktop width?

File: cypress/support/iterator.js
Comment (jamesbhobbs) [inline cypress/support/iterator.js:1]: Very nice

File: cypress/support/iterator.js
Comment (jamesbhobbs) [inline cypress/support/iterator.js]: This takes far too long to run to merge. We must add an smokeOnly arg which if true only runs against a hardcoded list of live services - we can refactor this later to not be hardcoded.

NB by doing this the default for all CI and CD it means an immediately following task  from this is a PR that runs all tests in jenkins - that is distinct from deploy so as not to slow merging. Ideally this would be triggered by each deploy or a Chron running at least once a day and ideally every 4-6 hours.

File: cypress/support/commands/pageTestFragment/cookie.js
Comment (jamesbhobbs) [inline cypress/support/commands/pageTestFragment/cookie.js]: Just repeat this between the cookie files, we don't need to factor out this kind of duplication in tests.

File: cypress/integration/specialFeatures/cookieBanner.canonical.js
Comment (jamesbhobbs) [inline cypress/integration/specialFeatures/cookieBanner.canonical.js]: why is this changed and not the amp variant?

**Review by jamesbhobbs — CHANGES_REQUESTED** [review]
Great work however this PR is far too big, I've left many comments, including some needing significant additions. I suggest this way forwards in separate PRs....

1. separate cookie changes
2. add iterator for a single page type in one PR (optionally add the smokeTest here)
3. Add the chron or deploy triggered E2E test run to jenkins
4. add iterator to all pagetypes with smokeTes enabled on jenkins and travis, and update the script for step 3 to not have smoke enabled.

File: cypress/support/bodyTestHelper.js
Comment (dr3) [reply cypress/support/bodyTestHelper.js:12]: >by testing the JSON sent with a page we avoid this case

Yeah, but means we cant share tests with amp and canonical

File: cypress/integration/pages/frontPage.js
Comment (dr3) [reply cypress/integration/pages/frontPage.js]: Error pages dont have a canonical

File: cypress/integration/pages/frontPage.js
Comment (dr3) [reply cypress/integration/pages/frontPage.js:10]: not sure dont wanna remove it in this pr though, there be dragons

File: cypress/support/bodyTestHelper.js
Comment (jamesbhobbs) [reply cypress/support/bodyTestHelper.js:12]: I think the likelihood of this kind of would be identical for AMP, do we already have this logic there? And if so have we ever seen this kind of error? If not then it seems safe.

File: cypress/integration/pages/frontPage.js
Comment (jamesbhobbs) [reply cypress/integration/pages/frontPage.js:10]: fair

File: cypress/integration/pages/frontPage.js
Comment (jamesbhobbs) [reply cypress/integration/pages/frontPage.js]: Good point hadn't thought of that. I think an if not errorpage would then make more sense given that isolated example, else we'll have to remember to put it everywhere and we shouldn't risk that on something so important.

File: cypress/integration/pages/index.js
Comment (jamesbhobbs) [inline cypress/integration/pages/index.js]:
```suggestion
      it('should render the BBC branding', () => {
```

File: cypress/integration/pages/errorPage404.js
Comment (jamesbhobbs) [inline cypress/integration/pages/errorPage404.js]:
```suggestion
      it(`should display a ${appConfig[service].translations.error[404].statusCode} error message on screen`, () => {
```

**Comment by jamesbhobbs** [comment]
Rereviewed missing is 

1. https://github.com/bbc/simorgh/pull/3053#discussion_r313619651
2. the ability to run this as a chron without smoke enabled. If this is merged as is we'd lose a lot of coverage on services not live or currently worked on and risk breaking them and not noticing for a long time. 

Re my stability concerns in https://github.com/bbc/simorgh/pull/3053#discussion_r313618969 I've rerun it circa 15 times and see no flake, so I think this is more stable than latest.

File: cypress/support/iterator.js
Comment (jamesbhobbs) [reply cypress/support/iterator.js]: NB to other CRers, this wasn't done 'add an smokeOnly arg which if true only runs against a hardcoded list of live services' and instead this was done: https://github.com/bbc/simorgh/pull/3123 which is more future proofed

**Comment by dr3** [comment]
> Rereviewed missing is
> 
> 1. [#3053 (comment)](https://github.com/bbc/simorgh/pull/3053#discussion_r313619651)
> 2. the ability to run this as a chron without smoke enabled. If this is merged as is we'd lose a lot of coverage on services not live or currently worked on and risk breaking them and not noticing for a long time.
> 
> Re my stability concerns in [#3053 (comment)](https://github.com/bbc/simorgh/pull/3053#discussion_r313618969) I've rerun it circa 15 times and see no flake, so I think this is more stable than latest.

@benjaminhobbs Just done #1, for #2 do you want that pre-merge of this, or is picking it up immediately after merge fine?

File: cypress/integration/application/index.js
Comment (jamesdonoh) [inline cypress/integration/application/index.js]: Can we reuse `serviceHasPageType` https://github.com/bbc/simorgh/pull/3053/files#diff-a372023abbaa311428537c61de200162R4?

File: cypress/support/iterator.js
Comment (jamesdonoh) [inline cypress/support/iterator.js]: I feel like the naming of these could make this easier to follow - do we mean `always` in the sense of tests that are run for both AMP and canonical pages? Yet they contain page type exclusions, for example https://github.com/bbc/simorgh/pull/3053/files#diff-6a133ac9ddaa5ea2078f57fc5c2567e1R25-R35. And `tests` doesn't really describe what these are.

Also how/why could `alwaysTests` be falsy?

File: cypress/support/iterator.js
Comment (jroebu14) [inline cypress/support/iterator.js:4]: I wrote the [exact same function](https://github.com/bbc/simorgh/blob/0a67dc52c6da8331d28261c2f7cf833b94d34578/cypress/support/serviceHasPageType/index.js) but it's curried like this:

```js
const serviceHasPageType = pageType => service =>
  pathSatisfies(
    Boolean,
    [service, 'pageTypes', pageType, 'path'],
    servicesConfig,
  );
```

enabling you to do this

```js
.filter(serviceHasPageType(pageType))
```

instead of this

```js
.filter(service => serviceHasPageType(service, pageType))
```

File: cypress/integration/pages/index.js
Comment (jroebu14) [inline cypress/integration/pages/index.js]: I'm not sure about the naming of this. Does it run common tests? Maybe something like `runCommonTests` would give a better hint to readers?

File: cypress/support/iterator.js
Comment (jroebu14) [inline cypress/support/iterator.js:7]: Couldn't tell what this function did where it was being used so had to read here and I'm still not sure

File: cypress/support/iterator.js
Comment (jroebu14) [inline cypress/support/iterator.js:41]: Could this file have a more meaningful name like `runTests` or `runTestsForPageType`?

File: cypress/integration/pages/frontPage.js
Comment (jroebu14) [inline cypress/integration/pages/frontPage.js]: since this function actually runs the tests when called what about naming it `runAmpTests` and same for canonical e.g. `runCanonicalTests`

**Review by jroebu14 — APPROVED** [review]
LGTM although I just had issues with the naming of some functions.

Also there's a tiny bit of overlap in one of my PRs. Might be worth having a look over and seeing if anything here is useful https://github.com/bbc/simorgh/pull/3084/files?w=1

Otherwise I'm happy to wait for this to be merged first so you don't have to deal with the conflicts.

**Review by jamesbhobbs — APPROVED** [review]

File: cypress/support/iterator.js
Comment (dr3) [reply cypress/support/iterator.js:4]: Gunna leave it as is, lower barrier for entry IMO

File: cypress/support/iterator.js
Comment (dr3) [reply cypress/support/iterator.js:41]: IMO iterator is a more apt name. @benjaminhobbs named it originally

File: cypress/support/iterator.js
Comment (dr3) [reply cypress/support/iterator.js:7]: It decides what pages are run during smoke test and what isnt, not sure of a new name, any ideas?

**Pull request merged**
